### PR TITLE
ch4: enablement on platforms with 8+ nics

### PR DESCRIPTION
**Pull request by wenduwan**
## Pull Request Description

Remove the 8-nic limit and enable MPICH to support high nic-count platforms

Ref: https://github.com/pmodels/mpich/issues/6688

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

File: src/mpid/ch4/netmod/ofi/ofi_nic.c [hunk 1]
```diff
@@ -181,12 +172,18 @@ static int setup_multi_nic(int nic_count)
...
     MPIR_CHKLMEM_DECL(1);
 
-    MPIDI_OFI_global.num_nics = nic_count;
-
     /* Initially sort the NICs by name. This way all intranode ranks have a consistent view. */
     qsort(MPIDI_OFI_global.prov_use, MPIDI_OFI_global.num_nics, sizeof(struct fi_info *),
           compare_nic_names);
 
+    if (nic_count > MPIDI_OFI_MAX_NICS) {
+        fprintf(stderr, "Warning: Detected %d NICs, but only %d(MPIDI_OFI_MAX_NICS) are considered.\n",
```
Comment (wenduwan) [inline src/mpid/ch4/netmod/ofi/ofi_nic.c:180]: Please advise a better way to print warnings.

**Comment by wenduwan** [comment]
@hzhou @raffenet Could you kindly review the change?

File: src/mpid/ch4/netmod/ofi/ofi_pre.h [hunk 2]
```diff
@@ -284,7 +284,7 @@ typedef struct {
 } MPIDI_OFI_win_t;
 
 /* Maximum number of network interfaces CH4 can support. */
-#define MPIDI_OFI_MAX_NICS 8
+#define MPIDI_OFI_MAX_NICS 128
```
Comment (hzhou) [inline src/mpid/ch4/netmod/ofi/ofi_pre.h:287]: I think we can remove this macro and rely on `MPIR_CVAR_CH4_OFI_MAX_NICS` to control the number of active NICs to use. I think we should reset `MPIR_CVAR_CH4_OFI_MAX_NICS` default to `8`.

Comment (wenduwan) [reply src/mpid/ch4/netmod/ofi/ofi_pre.h:287]: @hzhou Cool this is better. I can give it a go by next week.

I would also like second opinions from Aurora folks since they introduced the limit. Do you know a good contact?

Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_pre.h:287]: cc @zhenggb72 @sagarth

Comment (wenduwan) [reply src/mpid/ch4/netmod/ofi/ofi_pre.h:287]: @zhenggb72 @sagarth Some background on this PR. The 8 NIC limit is a blocker for AWS customers to use newer platforms with e.g. 32 NICs. We are exploring the possibility to raise or preferably remove the limit.

Do you see any downside of doing so?

**Comment by hzhou** [comment]
tag @nitbhat @tarudoodi  for review

File: src/mpid/ch4/netmod/ofi/ofi_pre.h (hunk 2 above)
Comment (tarudoodi) [reply src/mpid/ch4/netmod/ofi/ofi_pre.h:287]: I do like Hui's suggestion. We also went to keep the ability to change the number of NICs with a CVAR, in case we only want to use the NICs on first socket. Its an easier control.

@wenduwan Maybe you could set a machine specific value of this CVAR on your platforms?

Comment (wenduwan) [reply src/mpid/ch4/netmod/ofi/ofi_pre.h:287]: @hzhou After reading the usage of `MPIDI_OFI_MAX_NICS` closer, I think it requires a larger change to eliminate it.

I'm especially curious about these static counters https://github.com/pmodels/mpich/blob/69522c242b8106c1fa0dfa7fa52944eaadafbdde/src/mpid/ch4/netmod/ofi/globals.c#L10-L19

Do you think we can/should make them dynamically sized instead?

Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_pre.h:287]: Yes, I think they can be dynamically allocated

**Comment by wenduwan** [comment]
Converted change to draft for testing.

File: src/mpid/ch4/netmod/ofi/ofi_nic.c [hunk 3]
```diff
@@ -80,67 +79,83 @@ bool MPIDI_OFI_nic_already_used(const struct fi_info * prov, struct fi_info ** o
...
+#else
+        /* Always pick the first provider if NIC info is not available */
+        if (!first_prov) {
+            first_prov = fi_dupinfo(cur_prov);
+            first_prov->next = NULL;
+            ++nic_count;
+            break;
+        }
+#endif /* HAVE_LIBFABRIC_NIC */
+    }
```
Comment (wenduwan) [inline src/mpid/ch4/netmod/ofi/ofi_nic.c:123]: This change can be extracted to a separate commit if you like. I kept it in the big commit since `git add -p` does not allow me to split it out.

File: src/mpid/ch4/netmod/ofi/ofi_pre.h (hunk 2 above)
Comment (wenduwan) [reply src/mpid/ch4/netmod/ofi/ofi_pre.h]: @tarudoodi Apologies I missed your comment last week. I have retained the CVAR after the change. Please let me know if that is sufficient for your purpose.

File: src/mpid/ch4/netmod/ofi/ofi_pre.h [hunk 4]
```diff
@@ -283,15 +283,8 @@ typedef struct {
...
 
-/* Maximum number of network interfaces CH4 can support. */
-#define MPIDI_OFI_MAX_NICS 8
-
 typedef struct {
-#ifdef MPIDI_OFI_VNI_USE_DOMAIN
-    fi_addr_t dest[MPIDI_OFI_MAX_NICS][MPIDI_CH4_MAX_VCIS];     /* [nic][vci] */
-#else
-    fi_addr_t dest[MPIDI_OFI_MAX_NICS][1];
-#endif
```
Comment (wenduwan) [inline src/mpid/ch4/netmod/ofi/ofi_pre.h:294]: In my update I chose to ignore `MPIDI_OFI_VNI_USE_DOMAIN` - instead I made `dest` to be always `nics` x `MPIDI_CH4_MAX_VCIS`. I believe `MPIDI_CH4_MAX_VCIS` by default is 1.

**Comment by wenduwan** [comment]
@hzhou Please take a look at the update. I followed your suggestion to deprecate `MPIDI_OFI_MAX_NICS`. As a result I had to change a few static variables and data structures to make their size and memory dynamic. I tried to make sure the memory is alloced and freed properly.

For my testing, I ran the osu microbenchmarks on 2 different platforms with `./configure ... --enable-g=all`
1. hpc6a.48xlarge with 1 NIC
2. p5.48xlarge with 32 NICs

I added prints and verified that the code selects the right nic for each rank, and `MPIR_CVAR_CH4_OFI_MAX_NICS` can be used to restrict NIC visibility.

Please let me know your thoughts.

File: src/mpid/ch4/netmod/ofi/ofi_types.h [hunk 5]
```diff
@@ -399,7 +399,7 @@ typedef struct {
     /* Process management and PMI globals */
     int pname_set;
     int pname_len;
-    char addrname[MPIDI_OFI_MAX_NICS][FI_NAME_MAX];
+    char addrname[FI_NAME_MAX];
```
Comment (tarudoodi) [inline src/mpid/ch4/netmod/ofi/ofi_types.h:402]: Do we need to account for the new num_nics here too?

Comment (wenduwan) [reply src/mpid/ch4/netmod/ofi/ofi_types.h:402]: @tarudoodi I don't think so. This change is in its own commit. `addrname` is currently used as a singular char array/string, and I think it was defined as a 2-d array by mistake.

**Comment by tarudoodi** [comment]
@wenduwan The PR looks ok to me.  It keeps the usage that we had intended too. Please run the tests to make sure the testsuite passes.

**Comment by nitbhat** [comment]
> tag @nitbhat @tarudoodi for review

Sorry the delay in getting back on this. Reviewed, the changes look good. 

Do the CI tests exercise multi-nic code? Otherwise, it'll be good to manually test on some major multi-nic platforms to ensure that nic assignment looks good. (Platforms could be: aws instances, Aurora, Infiniband systems with multiple nics).

**Comment by wenduwan** [comment]
@hzhou Apologies for the inactivity. I brought up this change again to my leadership, and unfortunately we have decided to table the project for now due to resource contention. I will close the PR and reopen once we get a chance to re-evaluate the project.
